timelock-corruption-detection/src/test/java/com/palantir/timelock/history/PaxosLogHistoryProviderTest.java

```diff
@@ -0,0 +1,237 @@
+
+        List<HistoryQuery> historyQueries = ImmutableList.of(HistoryQuery.of(
+                ImmutableNamespaceAndUseCase.of(CLIENT, USE_CASE), lastVerified));
+        List<LogsForNamespaceAndUseCase> remoteHistory
+                = HistoryLoaderAndTransformer.getLogsForHistoryQueries(history, historyQueries);
+
+        when(remote.getPaxosHistory(any(), any())).thenReturn(PaxosHistoryOnRemote.of(remoteHistory));
+
+        List<CompletePaxosHistoryForNamespaceAndUseCase> completeHistory = paxosLogHistoryProvider.getHistory();
+
+        CompletePaxosHistoryForNamespaceAndUseCase historyForNamespaceAndUseCase
+                = Iterables.getOnlyElement(completeHistory);
+        assertSanityOfFetchedRecords(historyForNamespaceAndUseCase, CLIENT, USE_CASE, 100 - lastVerified);
+    }
+
+    @Test
+    public void canFetchAndCombineHistoriesAcrossNamespaces() {
+        List<HistoryQuery> historyQueries = new ArrayList<>();
+        Set<NamespaceAndUseCase> namespaceAndUseCases = new HashSet();
+
+        IntStream.rangeClosed(1, 100).forEach(idx -> {
+            String useCase = String.valueOf(idx);
+            Client client =  Client.of(useCase);
+
+            PaxosSerializationTestUtils.writeToLogs(
+                    createAcceptorLog(ImmutableNamespaceAndUseCase.of(client,
+                            AcceptorUseCase.createAcceptorUseCase(useCase).value())),
+                    createLearnerLog(ImmutableNamespaceAndUseCase.of(client,
+                            LearnerUseCase.createLearnerUseCase(useCase).value())),
+                    1, idx);
+
+            historyQueries.add(HistoryQuery.of(ImmutableNamespaceAndUseCase.of(client, useCase), -1));
+            namespaceAndUseCases.add(ImmutableNamespaceAndUseCase.of(client, useCase));
+        });
+
+        List<LogsForNamespaceAndUseCase> remoteHistory
+                = HistoryLoaderAndTransformer.getLogsForHistoryQueries(history, historyQueries);
+        when(remote.getPaxosHistory(any(), any())).thenReturn(PaxosHistoryOnRemote.of(remoteHistory));
+
+        List<CompletePaxosHistoryForNamespaceAndUseCase> completeHistory = paxosLogHistoryProvider.getHistory();
+        assertThat(completeHistory.size()).isEqualTo(100);
+
+        completeHistory.forEach(historyForNamespaceAndUseCase -> {
+            Client client = historyForNamespaceAndUseCase.namespace();
+            String useCase = historyForNamespaceAndUseCase.useCase();
+
+            assertSanityOfFetchedRecords(
+                    historyForNamespaceAndUseCase,
+                    client,
+                    useCase,
+                    getIntegerValueOfClient(historyForNamespaceAndUseCase));
+
+            // to assert that history for all clients was fetched
+            assertThat(namespaceAndUseCases.remove(ImmutableNamespaceAndUseCase.of(client, useCase))).isTrue();
+
+        });
+        assertThat(namespaceAndUseCases).isEmpty();
```

Comment:
Is this part really needed? I think it's probably more readable to just do a stream/map on completeHistory and check that against another stream rather than relying on a mutable set